### DISCUSSION TITLE
Background Task with websocket

### DISCUSSION THREAD

**Discussion by DrPyser** [fastapi/fastapi#8123] [Questions] [answered]
**Description**

Is it possible to use BackgroundTasks with websockets?

I tried with a simple case:

```python
def test():
   print('test')


@router.post("/test")
async def send_notification(background_tasks: BackgroundTasks):
   background_tasks.add_task(test)
   return {"message": "Notification sent in the background"}


@router.websocket("/ws")
async def listen(websocket: WebSocket, background_tasks: BackgroundTasks):
   await websocket.accept()
   background_tasks.add_task(test)
```
The `send_notification` http endpoint example works, but the `listen` websocket endpoint doesn't.

Thanks!

**Comment by tiangolo** [comment, marked as answer]
Background tasks internally depend on a `Request`, and they are executed after returning the request to the client. In the case of a WebSocket, as the WebSocket is not really "finished" but is an ongoing connection, it wouldn't be a background task.

Nevertheless, you can make sure an async function is executed with `asyncio.ensure_future`: https://docs.python.org/3/library/asyncio-future.html#asyncio.ensure_future

**Comment by DrPyser** [comment]
Okay, got it, thanks.

**Comment by tiangolo** [comment]
Cool! I'll close this issue now then.

**Comment by targhs** [comment]
@DrPyser Can you show how you did it?

**Comment by lokaimoma** [comment]
@targhs Here is an example. This came late but hope it helps someone.
![image](https://user-images.githubusercontent.com/46226886/154804431-91da01fe-6bf3-4415-bd0c-b25527412e8a.png)

**Comment by Nikita-Yasenkov** [comment]
@lokaimoma 
what's connection_manager in your case? Is it an external library?

**Reply by lokaimoma** [reply]
In my case, it's just a class I wrote to manage the different web socket connections. For instance sending a message to all connections or a group, etc

**Comment by flopana** [comment]
> @lokaimoma 
> what's connection_manager in your case? Is it an external library?
> 

The connection manager is most likely from the fast API documentation

https://fastapi.tiangolo.com/advanced/websockets/#handling-disconnections-and-multiple-clients
